pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -95,108 +153,152 @@ public static void init() {
-      throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
+    @Override
+    public PinotSqlFunction toPinotSqlFunction() {
+      return new PinotSqlFunction(_name, getReturnTypeInference(), getOperandTypeChecker());
     }
 
-    @ScalarFunction(names = {"jsonMatch", "json_match"}, isPlaceholder = true)
-    public static boolean jsonMatch(String text, String pattern) {
-      throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
+    private SqlReturnTypeInference getReturnTypeInference() {
+      return opBinding -> {
+        int numArguments = opBinding.getOperandCount();
+        FunctionInfo functionInfo = getFunctionInfo(numArguments);
+        Preconditions.checkState(functionInfo != null, "Failed to find function: %s with %s arguments", _name,
+            numArguments);
+        Method method = functionInfo.getMethod();
+        Class<?> returnClass = method.getReturnType();
+        RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        RelDataType returnType = returnClass == Object.class ? typeFactory.createSqlType(SqlTypeName.ANY)
+            : JavaTypeFactoryImpl.toSql(typeFactory, typeFactory.createJavaType(returnClass));
+
+        if (!functionInfo.hasNullableParameters()) {
+          // When any parameter is null, return is null
+          for (RelDataType type : opBinding.collectOperandTypes()) {
+            if (type.isNullable()) {
+              return typeFactory.createTypeWithNullability(returnType, true);
+            }
+          }
+        }
+
+        return method.isAnnotationPresent(Nullable.class) ? typeFactory.createTypeWithNullability(returnType, true)
+            : returnType;
+      };
+    }
+
+    private SqlOperandTypeChecker getOperandTypeChecker() {
+      if (_functionInfoMap.containsKey(VAR_ARG_KEY)) {
+        return OperandTypes.VARIADIC;
+      }
+      if (_functionInfoMap.size() == 1) {
+        return getOperandTypeChecker(_functionInfoMap.values().iterator().next().getMethod());
+      }
+      List<SqlOperandTypeChecker> operandTypeCheckers = new ArrayList<>(_functionInfoMap.size());
+      for (FunctionInfo functionInfo : _functionInfoMap.values()) {
+        operandTypeCheckers.add(getOperandTypeChecker(functionInfo.getMethod()));
+      }
+      return OperandTypes.or(operandTypeCheckers.toArray(new SqlOperandTypeChecker[0]));
+    }
+
+    private static SqlTypeFamily getSqlTypeFamily(Class<?> clazz) {
+      // NOTE: Pinot allows some non-standard type conversions such as Timestamp <-> long, boolean <-> int etc. Do not
+      //       restrict the type family for now. We only restrict the type family for String so that cast can be added.
+      //       Explicit cast is required to correctly convert boolean and Timestamp to String.
+      // TODO: Revisit this.
+      return clazz == String.class ? SqlTypeFamily.CHARACTER : SqlTypeFamily.ANY;
```

Comment:
In v1, we are handling majority of the type casting implicitly. Currently we need explicit casting from BOOLEAN and TIMESTAMP to STRING, or they will be converted with internal type (e.g. BOOLEAN will be case to "1" or "0" instead of "true" or "false"). Added the example to the javadoc